### ISSUE TITLE
Directional Derivatives - FMUs for Co-Simulation

### ISSUE THREAD

**Issue by ig-borges** [OpenModelica/OpenModelica#7371]
Hi all,

I would like to know if there is a possibility to export FMUs, providing directional derivatives, and if possible, how can I enable this option when exporting an FMU.

Thanks!

**Comment by lochel** [comment]
OpenModelica does only support directional derivatives for model exchange FMUs at the moment. But you are asking specifically for co-simulation FMUs, right?

**Comment by ig-borges** [comment]
Yes. I would like to export CS-type FMUs providing directional derivatives. I'm doing some co-simulation tests using PyFMI, and when I use the function 'get_capability_flags ( )' for FMUs, I see 'providesDirectionalDerivatives': False.

**Comment by lochel** [comment]
Thanks for requesting this feature. It is on our list now.

**Comment by awecefil** [comment]
Hi @lochel , could you tell me how to export ME-type FMU with providing directional derivatives? I have read this [document](https://openmodelica.org/doc/OpenModelicaUsersGuide/OpenModelicaUsersGuide-latest.pdf) and add `-d=disableDirectionalDerivatives` in Additional Translation Flag, but the modelDescription.xml of exported FMU only shows `providesDirectionalDerivative="true`" for CS-type.

Besides, is it possible to export ME-type FMU with `canGetAndSetFMUstate="true"`? I also tried adding `-d=fmiExperimental`, but it also only enable it for CS-type.

**Comment by casella** [comment]
@arun3688 can you give some feedback on that?

**Comment by arun3688** [comment]
@awecefil the issue is fixed with this commit https://github.com/OpenModelica/OpenModelica/commit/021284b4547b3c956408de08b07f693144539445, please test it from the nightly builds tomorrow

**Comment by jschueller** [comment]
- SimCodeUtil.providesDirectionalDerivative checks FmiModelStructure.
  continuousPartialDerivatives, which is always NONE() — the code generator
  never populates it.
- The Jacobian callbacks (initialPartialFMIDER, functionJacFMIDER_column) are
  never generated for FMUs.
- Without the Jacobian, _has_jacobian stays 0 and the runtime returns
  fmi2Discard/fmi2Error.

So the backend never computes or generates the symbolic Jacobian for FMU
export. That's the core missing piece.
